tests/test_secondary.py

```diff
@@ -0,0 +1,721 @@
+    """
+    Tests uptane.clients.secondary.Secondary::__init__()
+    Note that this doesn't test the root files provided to the constructor, as
+    those aren't used at all in the initialization; those will be tested by
+    attempting an update in the test for process_metadata below.
+    """
+    # Initialization should be done here, in the test of __init__, not earlier,
+    # and we will use the clients set up here in further testing (rather than
+    # creating a new client/clients in every single test).
+    # Consequently, these variables are stuck being set here, and so are stuck
+    # as globals.
+    global secondary_instance_1
+    global secondary_instance_2
+    global secondary_instance_3
+
+    # TODO: Test with invalid pinning file
+    # TODO: Test with pinning file lacking a Director repo.
+
+    # Now try creating a Secondary with a series of bad arguments, expecting
+    # errors.
+
+    # Invalid VIN:
+    with self.assertRaises(tuf.FormatError):
+      s = secondary.Secondary(
+          full_client_dir=TEMP_CLIENT_DIR_1,
+          director_repo_name=demo.DIRECTOR_REPO_NAME,
+          vin=5,
+          ecu_serial=ecu_serial1,
+          ecu_key=secondary_ecu_key,
+          time=clock,
+          timeserver_public_key=key_timeserver_pub,
+          firmware_fileinfo=factory_firmware_fileinfo,
+          director_public_key=None,
+          partial_verifying=False)
+
+    # Invalid ECU Serial
+    with self.assertRaises(tuf.FormatError):
+      s = secondary.Secondary(
+          full_client_dir=TEMP_CLIENT_DIR_1,
+          director_repo_name=demo.DIRECTOR_REPO_NAME,
+          vin=vin1,
+          ecu_serial=500,
+          ecu_key=secondary_ecu_key,
+          time=clock,
+          timeserver_public_key=key_timeserver_pub,
+          firmware_fileinfo=factory_firmware_fileinfo,
+          director_public_key=None,
+          partial_verifying=False)
+
+    # Invalid ECU Key
+      s = secondary.Secondary(
+          full_client_dir=TEMP_CLIENT_DIR_1,
+          director_repo_name=demo.DIRECTOR_REPO_NAME,
+          vin=vin1,
+          ecu_serial=ecu_serial1,
+          ecu_key={''},
+          time=clock,
+          timeserver_public_key=key_timeserver_pub,
+          firmware_fileinfo=firmware_fileinfo,
+          director_public_key=None,
+          partial_verifying=False)
+
+    # Invalid time:
+    with self.assertRaises(tuf.FormatError):
+      s = secondary.Secondary(
+          full_client_dir=TEMP_CLIENT_DIR_1,
+          director_repo_name=demo.DIRECTOR_REPO_NAME,
+          vin=vin1,
+          ecu_serial=ecu_serial1,
+          ecu_key=secondary_ecu_key,
+          time='potato',
+          timeserver_public_key=key_timeserver_pub,
+          firmware_fileinfo=factory_firmware_fileinfo,
+          director_public_key=key_directortargets_pub,
+          partial_verifying=False)
+
+    # Invalid director_public_key:
+    with self.assertRaises(tuf.FormatError):
+      s = secondary.Secondary(
```

Comment:
I suppose it doesn't matter, since these are all expected to raise exceptions. I'll remove the assignment.